File: frontend/code-editor-fe/src/app/services/file.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class FileService {

  private baseUrl = 'http://localhost:8080/api/files';
  
  constructor(private http: HttpClient) {}
  
  getFilesByRoom(roomId: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/room/${roomId}`);
  }
  
  createFile(roomId: string, request: any): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/room/${roomId}`, request);
  }
  
  updateFile(fileId: number, request: any): Observable<any> {
    return this.http.put<any>(`${this.baseUrl}/${fileId}`, request);
  }

  deleteFile(fileId: number): Observable<any> {
    return this.http.delete<any>(`${this.baseUrl}/${fileId}`);
  }

  getSnapshot(roomId: string, fileId: number) {
  return this.http.get<{ roomId: string; fileId: number; content: string; revision: number }>(
    `http://localhost:8080/api/collab/rooms/${roomId}/files/${fileId}/snapshot`
  );
}
  
  private getLanguageFromFileName(fileName: string): string {
    const extension = fileName.split('.').pop()?.toLowerCase();
    const languageMap: { [key: string]: string } = {
      // C++
      'cpp': 'cpp',
      'cc': 'cpp',
      'cxx': 'cpp',
      'c++': 'cpp',
      
      // Java
      'java': 'java',
      
      // Python
      'py': 'python',
      
      // JavaScript
      'js': 'javascript',
      'jsx': 'javascript',
      
      // HTML
      'html': 'html',
      'htm': 'html',
      
      // CSS
      'css': 'css'
    };
    
    return languageMap[extension || ''] || 'plaintext';
  }
}

File: backend/backend-service/src/main/java/com/example/backend_service/services/FileService.java
package com.example.backend_service.services;

import com.example.backend_service.dto.CreateFileRequest;
import com.example.backend_service.dto.UpdateFileRequest;
import com.example.backend_service.entity.EditorFile;
import com.example.backend_service.entity.EditorRoom;
import com.example.backend_service.exception.ResourceNotFoundException;
import com.example.backend_service.repository.EditorFileRepository;
import com.example.backend_service.repository.EditorRoomRepository;
import java.util.List;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
public class FileService {

  @Autowired
  private EditorFileRepository fileRepository;

  @Autowired
  private EditorRoomRepository roomRepository;

  public EditorFile createFile(String roomId, CreateFileRequest request) {
    // Check if room exists
    EditorRoom room = roomRepository.findByRoomId(roomId);
    if (room == null) {
      throw new ResourceNotFoundException("Room not found with ID: " + roomId);
    }

    // Check if file with same name already exists in the room
    EditorFile existingFile = fileRepository.findByRoomIdAndFileName(roomId, request.getFileName());
    if (existingFile != null) {
      throw new IllegalArgumentException(
          "File with name '" + request.getFileName() + "' already exists in this room");
    }

    // Determine language from file extension if not provided
    String language = request.getLanguage();
    if (language == null || language.trim().isEmpty()) {
      language = determineLanguageFromFileName(request.getFileName());
    }

    // Create new file
    EditorFile file = new EditorFile(
        request.getFileName(),
        request.getContent() != null ? request.getContent() : "",
        language
    );
    file.setRoom(room);
    return fileRepository.save(file);
  }

  @Transactional(readOnly = true)
  public List<EditorFile> getFilesByRoom(String roomId) {
    // Verify room exists
    if (!roomRepository.existsByRoomId(roomId)) {
      throw new ResourceNotFoundException("Room not found with ID: " + roomId);
    }

    return fileRepository.findByRoomId(roomId);
  }

  public EditorFile updateFile(Long fileId, UpdateFileRequest request) {
    EditorFile file = fileRepository.findById(fileId)
        .orElseThrow(() -> new ResourceNotFoundException("File not found with ID: " + fileId));

    // Update content
    file.setContent(request.getContent());

    // Update language if provided
    if (request.getLanguage() != null && !request.getLanguage().trim().isEmpty()) {
      file.setLanguage(request.getLanguage());
    }

    return fileRepository.save(file);
  }

  @Transactional(readOnly = true)
  public EditorFile getFileById(Long fileId) {
    return fileRepository.findById(fileId)
        .orElseThrow(() -> new ResourceNotFoundException("File not found with ID: " + fileId));
  }

  public void deleteFile(Long fileId) {
    EditorFile file = getFileById(fileId);
    fileRepository.delete(file);
  }

  // Helper method to determine language from file extension
  private String determineLanguageFromFileName(String fileName) {
    if (fileName == null || !fileName.contains(".")) {
      return "plaintext";
    }

    String extension = fileName.substring(fileName.lastIndexOf(".") + 1).toLowerCase();

    switch (extension) {
      // C++
      case "cpp":
      case "cc":
      case "cxx":
      case "c++":
        return "cpp";

      // Java
      case "java":
        return "java";

      // Python
      case "py":
        return "python";

      // JavaScript
      case "js":
      case "jsx":
        return "javascript";

      // HTML
      case "html":
      case "htm":
        return "html";

      // CSS
      case "css":
        return "css";

      default:
        return "plaintext";
    }
  }
}
